### PR TITLE
Include init containers in error messages

### LINKED ISSUE #27040: kubectl logs UX when init containers are used
Im extracting logs part from #25818 

The current behavior is as follow: 

- when pod has one container
`kubectl get logs` return logs for this one container

- when pod has more than one container
`kubectl get logs` returns error:
`Error from server: a container name must be specified for pod kube-dns-v14-z0ho6, choose one of: [kubedns dnsmasq healthz]
`
As you can see in both cases init containers are ignored. The question are:

-  in first case should we return error and list all containers including init containers? 
- in second case should we also list init containers?

### PR DESCRIPTION
**Pull request by lukaszo**
fixes: #27040

### CODE REVIEW

File: pkg/registry/pod/strategy_test.go [hunk 1]
```diff
@@ -199,6 +199,22 @@ func TestCheckLogLocation(t *testing.T) {
...
 						{Name: "container2"},
 					},
+					InitContainers: []api.Container{
+						{Name: "initcontainer1"},
+					},
+				},
+				Status: api.PodStatus{},
+			},
+			opts:        &api.PodLogOptions{},
+			expectedErr: errors.NewBadRequest("a container name must be specified for pod test, choose one of: [container1 container2] or one of init containers: [initcontainer1]"),
```
Comment (dims) [inline pkg/registry/pod/strategy_test.go]: "or one of the init containers:" probably sounds slightly better

File: pkg/registry/pod/strategy.go [hunk 2]
```diff
@@ -424,8 +429,13 @@ func streamLocation(
 		case 0:
 			return nil, nil, errors.NewBadRequest(fmt.Sprintf("a container name must be specified for pod %s", name))
 		default:
-			containerNames := getContainerNames(pod)
-			return nil, nil, errors.NewBadRequest(fmt.Sprintf("a container name must be specified for pod %s, choose one of: [%s]", name, containerNames))
+			containerNames := getContainerNames(pod.Spec.Containers)
+			initContainerNames := getContainerNames(pod.Spec.InitContainers)
+			err := fmt.Sprintf("a container name must be specified for pod %s, choose one of: [%s]", name, containerNames)
+			if len(initContainerNames) > 0 {
+				err += fmt.Sprintf(", or one of init containers: [%s]", initContainerNames)
```
Comment (dims) [inline pkg/registry/pod/strategy.go]: "or one of the init containers:" probably sounds slightly better

Comment (lukaszo) [inline pkg/registry/pod/strategy.go]: ok, fixing

**Comment by lukaszo** [comment]
@smarterclayton @dchen1107 can you review?

**Comment by smarterclayton** [comment]
LGTM but doesn't make the bar for 1.3 right now.

**Comment by lukaszo** [comment]
@k8s-bot test this github issue: #27276

**Comment by sttts** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by fejta** [comment]
@k8s-bot unit test this issue: #28076

**Comment by fejta** [comment]
@k8s-bot unit test this issue: #28076

**Comment by sttts** [comment]
@k8s-bot unit test this issue: #28076

**Pull request merged**